sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/ApproximateDistinct.java

```diff
@@ -0,0 +1,600 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sketching;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.clearspring.analytics.stream.cardinality.CardinalityMergeException;
+import com.clearspring.analytics.stream.cardinality.HyperLogLogPlus;
+import com.google.auto.value.AutoValue;
+import com.google.common.base.MoreObjects;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import javax.annotation.Nullable;
+
+import org.apache.beam.sdk.coders.ByteArrayCoder;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.coders.CustomCoder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.transforms.Combine;
+import org.apache.beam.sdk.transforms.Combine.CombineFn;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+
+/**
+ * {@link PTransform}s for computing the approximate number of distinct elements in a stream.
+ * <br>The {@link CombineFn} is also exposed directly so it can be composed or used as a state cell
+ * in a stateful {@link ParDo}. See {@link ApproximateDistinctFn}.
+ *
+ * <p>This class relies on the HyperLogLog algorithm, and more precisely HyperLogLog+,
+ * the improved version of Google.
+ *
+ * <h2>References</h2>
+ *
+ * <p>The implementation comes from Addthis' Stream-lib library :
+ * <a>https://github.com/addthis/stream-lib</a>
+ * <br>The original paper of the HyperLogLog is available here :
+ * <a>http://algo.inria.fr/flajolet/Publications/FlFuGaMe07.pdf</a>
+ * <br>A paper from the same authors to have a clearer view of the algorithm is available here :
+ * <a>http://cscubs.cs.uni-bonn.de/2016/proceedings/paper-03.pdf</a>
+ * <br>Google's HyperLogLog+ version is detailed in this paper :
+ * <a>https://research.google.com/pubs/pub40671.html</a>
+ *
+ * <p>Three parameters can be tuned according to the processing context :
+ * <ul>
+ *    <li> Input Coder : The coder of the objects to combine.
```

Comment:
Note that this is only relevant for the combine fn. The transform should get the input coder from the input PCollection's .getCoder(). This also applies to a few other points in the documentation below that mention the coder.